README.md
Caddis
---
> On-The-Fly Mock JSON Server

![caddis npm](https://nodei.co/npm/caddis.png)

Installation
---
```
$ npm install -g caddis
```

```
$ caddis -h

usage: caddis [action] [port]

Starts a server at http://localhost:<port> as a daemon, exposing an api to post JSON to in order to mock a RESTful service.
Defaults to port 3001 unless --port option provided.

actions:
  start               Start Caddis at http://localhost:<port>
  stop                Stop a previously started Caddis daemon
options:
  -h                  Display this help menu
  -p                  Desired port to start server on localhost
```

Usage
---
```
$ caddis start
$ curl -X POST -d '{"method":"GET", "uri":"/foo", "response":{"bar":"baz"}}' http://localhost:3001/api --header "Content-Type:application/json"
```

Visit [http://localhost:3001/foo](http://localhost:3001/foo), prints:

```
{
  bar: 'baz'
}
```

```
$ caddis stop
```

Visit [http://localhost:3001/foo](http://localhost:3001/foo), Not Found.

What?
---
![Caddis Fly Lure](http://custardbelly.com/images/caddis.jpg)  
A caddis is a moth-like insect often used as models for fly lures in fishing.

The [caddis](https://github.com/bustardcelly/caddis) CLI tool is used to start and stop a RESTful JSON service with the ability to POST route configuration and responses, on-the-fly, for mocking and testing purposes.

_It may be a stretch, but there's wit in there somewhere..._

Why?
---
There are other projects I have been a part of, such as [madmin](https://github.com/infrared5/madmin), that allow for dynamically creating RESTful APIs through a User Interface and allows for persistance through I/O.

Recently, I was involved with mocking a service layer for unit testing purposes and found that the manual curation of such an API was too tedious for the task at hand - I wanted the process to be much more fluid and simple.

In this particular instance I needed to:

1. Start a server in setup/before
2. Dynamically add a route to the service with mock JSON response
3. Run the test
4. Shut down the server in teardown/after

Fairly simple, and most of all I didnt want any artifacts lying around - in other words I didn't need for any routes that I dynamically created to stick around on my local disk after the tests were done.

As such, [caddis](https://github.com/bustardcelly/caddis) was born.

How
---
As mentioned briefly above, [caddis](https://github.com/bustardcelly/caddis) is a CLI tool. It is recommended to install globally:

```
$ npm install -g caddis
```
_: you may need to `sudo`_

Once installed, you can start the service (currently defaults to [http://localhost:3001](http://localhost:3001)) and begin POSTing route configurations in JSON. Here is an example using cUrl that dynamically adds a GET route at `/foo` with a simple JSON payload of `{"bar":"baz"}`:

```
$ curl -X POST -d '{"method":"GET", "uri":"/foo", "response":{"bar":"baz"}}' http://localhost:3001/api --header "Content-Type:application/json"
```

You are not confiuned to cUrl - you can use whatever networking library in whatever language you are writing your tests in and the server can handle all modern RESTful methods:

* GET
* POST
* PUT
* DELETE

When you are finished, simply stop the [caddis](https://github.com/bustardcelly/caddis) server:

```
$ caddis stop
```

## Post Data
The post body data that is sent to [caddis](https://github.com/bustardcelly/caddis) has the following properties:

* __method__: The REST method
* __uri__: The endpoint of the service to hit (requires prepended '/')
* __response__: The JSON object to return on request
* __statusCode__: [Optional] Status code to return. Defaults to 200
* __delay__: [Optional] Time delay, in milliseconds, to delay response. Defaults to NaN

Tests
---

```
$ npm run test
```

License
---
Copyright (c) 2014 Todd Anderson

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.


HISTORY.md
0.4.4 / 2014-03-13
==================
  * allow special characters in qury parameter values.

0.4.3 / 2014-03-13
==================
  * support for tokenized query parameters.
  * spec for multuple param uri.
  * added help notification to README

0.4.2 / 2014-03-04
==================
  * allow for special characters in uri param values.

0.4.1 / 2014-03-03
==================
  * support for return of call history through /_call_requests. closes https://github.com/bustardcelly/caddis/issues/4

0.4.0 / 2014-03-03
==================
  * lightweight body parser to accept param string and JSON.
  * perhaps some support for request param parse on body.
  * support for port option!
  * proper termination on regex matching for uri on add route.
  * update to forever runner on caddis to preseved session-only logging at HOME for developer debugging purposes.
  * light logging in caddis server.
  * experimental removal of express to allow custom ports.
  * up of node version dependency.

0.3.6 / 2014-02-26
==================
  * delay response support. closes [#3](https://github.com/bustardcelly/caddis/issues/3)

lib/caddis/server.js
'use strict';

module.exports = function (port, options, callback) {
  if (!callback) {
    callback = options;
    options = {};
  }

  if (!options) options = {};

  var http = require('http');

  var routeMap = {};
  var routeSupport = ['get', 'post', 'put', 'delete'];
  var callHistory = [];

  var ROOT_ENDPOINT = /^\/$/i;
  var API_ENDPOINT = /\/api/i;
  var CALL_HISTORY_ENDPOINT = /\/_call_requests/i;
  var COLON_TOKEN = /(:[^:\/]+)[^\/]?/gi;
  var COLON_REPLACE = '[^\/]+';
  var QUERY_PARAM_TOKEN = /[\\?&]*=([^&]*)?/gi;
  var QUERY_PARAM_REPLACE = '=[^&]+';

  var print = function(message) {
    if (options.enableLogging) {
      console.log([new Date(), message].join(':: '));
    }
  };

  var requestIsRoot = function(req) {
    return req.length === 0 || req.url.match(ROOT_ENDPOINT);
  };

  var requestIsHistory = function(req) {
    return req.length === 0 || req.url.match(CALL_HISTORY_ENDPOINT);
  };

  var requestIsAPIRoute = function(req) {
    return req.method.toLowerCase() === 'post' && req.url.match(API_ENDPOINT);
  };

  var isValidInt = function(value) {
    var n = parseInt(value, 10);
    return !isNaN(n) && value % 1 === 0;
  };

  var validateStatusCode = function(value) {
    // Enhancement: Validate as proper HTTP status code.
    return isValidInt(value) ? parseInt(value, 10) : Number.NaN;
  };

  var validateDelay = function(value) {
    // Enhancement: Validate as proper timespan.
    return isValidInt(value) ? parseInt(value, 10) : Number.NaN;
  };

  var tokenizeURI = function(uri) {
    uri = uri.replace(COLON_TOKEN, COLON_REPLACE);
    uri = uri.replace('?', '\\?');
    uri = uri.replace(QUERY_PARAM_TOKEN, QUERY_PARAM_REPLACE);
    return '^' + uri.replace(COLON_TOKEN, COLON_REPLACE) + '$';
  };

  var addRoute = function(route) {
    var normalizedMethod = route.method.toLowerCase();
    if(routeMap.hasOwnProperty(normalizedMethod)) {
      route.regexStr = tokenizeURI(route.uri);
      route.regex = new RegExp(route.regexStr);
      route.match = function(uri) {
        print('Attempt to match: ' + this.regexStr + ', ' + uri);
        return uri.match(this.regex);
      };
      routeMap[normalizedMethod].push(route);
    }
  };

  var addToCallHistory = function(request, response) {
    var callEntry = {
      request: (typeof request === 'string') ? JSON.parse(request) : request,
      response: (typeof response === 'string') ? JSON.parse(response) : response
    };
    callHistory.push(callEntry);
    print('Added to history: ' + JSON.stringify(callEntry, null, 2));
  };

  var findRouteFromRequest = function(req) {
    var routeList = routeMap[req.method.toLowerCase()];
    var route;
    var index;
    if(routeList) {
      index = routeList.length;
      while(--index > -1) {
        route = routeList[index];
        if(route.match(req.url)) {
          return route;
        }
      }
    }
    return undefined;
  };

  // Solution coutesy of:
  // http://stackoverflow.com/questions/8648892/convert-url-parameters-to-a-javascript-object
  var bodyParser = function(value) {
    var data = value;
    // Check if issuer passed as params.
    if(!data.match(/^{(.*)/)) {
      print('Recieved response payload as param string.');
      data = JSON.parse('{"' + value.replace(/"/g, '\\"').replace(/&/g, '","').replace(/\=/g,'":"') + '"}');
      data.response = JSON.parse(data.response.replace(/\\"/g, '"'));
      print('Parsed and formatted to ' + JSON.stringify(data, null, 2));
    }
    // If not, assuming as json string.
    else {
      print('Received response payload as proper JSON string.');
      data = JSON.parse(data);
      print('Parsed and formatted as: ' + JSON.stringify(data, null, 2));
    }
    return data;
  };

  var establishRouteMap = function() {
    while(routeSupport.length > 0) {
      routeMap[routeSupport.shift()] = [];
    }
  };

  var server = http.createServer(function(request, response) {
    var data = '';
    var body;

    // Cross Domain.
    response.setHeader('Access-Control-Allow-Origin', '*');
    response.setHeader('Access-Control-Allow-Methods', 'GET,PUT,POST,DELETE');
    response.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
    request.setEncoding('utf8');
    request.on('data', function(chunk) {
      data += chunk;
    });

    if(request.method === 'OPTIONS') {
      request.on('end', function() {
        response.writeHead(200);
        response.end('\n');
        print('OPTIONS request.');
      });
    }
    else if(requestIsRoot(request)) {
      request.on('end', function() {
        body = JSON.stringify({result: 'hello, caddis!'}) + '\n';
        response.writeHead(200, {
          'Content-Type': 'application/json',
          'Content-Length': body.length
        });
        response.end(body);
        print('Root request.');
      });
    }
    else if(requestIsHistory(request)) {
      request.on('end', function() {
        body = JSON.stringify(callHistory) + '\n';
        response.writeHead(200, {
          'Content-Type': 'application/json',
          'Content-Length': body.length
        });
        response.end(body);
        print('Call history request.');
        print('Response: ' + body);
      });
    }
    else if(requestIsAPIRoute(request)) {
      request.on('end', function() {
        var route = bodyParser(decodeURIComponent(data));
        addRoute(route);
        body = JSON.stringify({result: true}) + '\n';
        response.writeHead(200, {
          'Content-Type': 'application/json',
          'Content-Length': body.length
        });
        response.end(body);
        print('Route added to ' + route.method + '.');
        print(JSON.stringify(route, null, 2));
      });
    }
    else {
      request.on('end', function() {
        var route;
        var delay;
        var statusCode;
        var timeout;
        var respond;

        route = findRouteFromRequest(request);
        if(route !== undefined) {
          delay = validateDelay(route.delay);
          statusCode = validateStatusCode(route.status);
          body = JSON.stringify(route.response) + '\n';
          respond = function() {
            response.writeHead(!isNaN(statusCode) ? statusCode : 200, {
              'Content-Type': 'application/json',
              'Content-Length': body.length
            });
            response.end(body);

            addToCallHistory({
              method: request.method,
              url: request.url,
              statusCode: request.statusCode ? request.statusCode : 200,
              body: data && data.length > 0 ? data : null
            }, body);
          };
          print('Stored route found.');
          print('Returning: ' + body +
                      'after ' + (!isNaN(delay) ? delay : 0) +
                      ' milliseconds with status ' + (!isNaN(statusCode) ? statusCode : 200));
          if(!isNaN(delay)) {
            timeout = setTimeout(function() {
              clearTimeout(timeout);
              respond();
            }, delay);
          }
          else {
            respond();
          }
        }
        else {
          body = JSON.stringify({
            message: request.url + ' not implemented.'
          }) + '\n';
          response.writeHead(501, { 
            'Content-Type': 'application/json',
            'Content-Length': body.length
          });
          response.end(body);
        }
      });
    }
  });

  establishRouteMap();
  server.listen(port, callback);

  print(new Date() + ':: Caddis started on port ' + port + '.');
};

if (require.main === module) {
  var port = 3001;
  module.exports(port, { enableLogging: true }, function() {});
}
